source/rust_verify/tests/bitvector.rs

```diff
@@ -0,0 +1,95 @@
+#![feature(rustc_private)]
+#[macro_use]
+mod common;
+use common::*;
+
+test_verify_one_file! {
+    #[test] test1 code! {
+        #[proof]
+        fn test1(b: u32) {
+            assert_bit_vector(b & 7 == b % 8);
+            assert(b & 7 == b % 8);
+
+            assert_bit_vector(b ^ b == 0);
+            assert(b ^ b == 0);
+
+            assert_bit_vector(b & b == b);
+            assert(b & b == b);
+
+            assert_bit_vector(b & 0 == 0);
+            assert(b & 0 == 0);
+
+            assert_bit_vector(b | b == b);
+            assert(b | b == b);
+        }
+    } => Ok(())
+}
+
+test_verify_one_file! {
+    #[test] test2 code! {
+        #[proof]
+        fn test2(b: u32) {
+            assert_bit_vector(b & 0xff < 0x100);
+            assert(b & 0xff < 0x100);
+            //assert(0xff & b < 0x100);  // fails without communtativity
+
+            assert_bit_vector(b<<2 == b*4);
+            assert(b<<2 == b*4);
+
+            assert_bit_vector(b>>1 == b/2);
+            assert(b>>1 == b/2);
+        }
+    } => Ok(())
+}
+
+test_verify_one_file! {
+    #[test] test3 code! {
+        #[proof]
+        fn test3(b: u32) {
+            assert_bit_vector(2*b - b == b);
+            assert(2*b - b == b);
+
+            assert_bit_vector(b <= b);
+            assert_bit_vector(b >= b);
+
+            assert_bit_vector(b & b & b == b | b | (b ^ b));
+            assert(b & b & b == b | b | (b ^ b));
+        }
+    } => Ok(())
+}
+
+test_verify_one_file! {
+    #[test] test1_fails code! {
+        #[proof]
+        fn test1(b: u32) {
+            assert_bit_vector(b | b > b); // FAILS
+        }
+    } => Err(err) => assert_one_fails(err)
+}
+
+test_verify_one_file! {
+    #[test] test2_fails code! {
+        #[proof]
+        fn test2(b: u32) {
+            assert_bit_vector(b + 1 == b); // FAILS
+        }
+    } => Err(err) => assert_one_fails(err)
+}
+
+test_verify_one_file! {
+    #[test] test3_fails code! {
+        #[proof]
+        fn test3(b: u32) {
+            assert_bit_vector(b & 0 > 0); // FAILS
+        }
+    } => Err(err) => assert_one_fails(err)
+}
+
+test_verify_one_file! {
+    #[test] test4_fails code! {
+        #[proof]
+        fn test4(b: u32) {
+            assert_bit_vector( (b << 2) >> 2 == b); // FAILS
+        }
+    } => Err(err) => assert_one_fails(err)
+}
```

Comment:
It would be good to add a failing test where the `assert_bit_vector` succeeds but the subsequent `assert` about integers fails (because it doesn't match the `assert_bit_vector`).

It might also be good to add a failing test to check that signed integers don't accidentally use the unsigned integer reasoning.

Reply:
I agree that those tests are great to have. 
About banning signed integers, I will make a few more commits to make sure of that before merging this into main.